### ISSUE TITLE
  what():  Type is not convertible to string

### ISSUE THREAD

**Issue by unixneo** [godlikemouse/ForceDirectedLayout#3]
Hey Jason!

On a large  data set we are working on with over 100,000 alert processed to around 7,500 node and 7,500 edges, , processing a JSON to an XML file (3D), we get this error caused by some data:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Type is not convertible to string
```

I've tried debugging, but the data set is more than 3MB and  the error from **fdl** does not provide a line number  or an attribute of the JSON data causing the problem.

Any clues on how to track down this error in fdl?

I've tried filtering the data to remove special chars, etc but that did not work so I might have been filtering the wrong data.  

Any idea which line in your code would throw this error when processing a JSON file?

`what():  Type is not convertible to string`

**Comment by godlikemouse** [comment]
Hi unixneo,

One of the first things I would do when attempting to debug a JSON issue like this, is to use a JSON validator to ensure the JSON is correct.  Next, I would remove all the links from the JSON and try again to see if the problem exists in the nodes or the edges.  Next, I would begin removing the first half of where the problem is (nodes or edges) and keep going from there.  Each time something works you know you're getting closer to the issue.  It sounds like an invalid attribute "Type" has been specified somewhere.

**Comment by unixneo** [comment]
Hey Jason,

Just a quick update.   I examined the input JSON file which was around 20,000 nodes and edges; and after hours of tracing work,  traced the problem to a "<" char in a string which was encapsulated by quotes and also an int attribute which did not have quotes.

Seems the fdl code woud be a bit more robust if it would accept integers and basic special chars as attributes and just them convert them to strings and process (pass) them through.   

Huge data sets will often have attributes as special chars like a "<" sign; and integers are very common as network attributes.   This great fdl code could be a bit more friendly in the "typing" area and accept these strings as "strings".   

Also,  when fdl throws a "typing" error it would help if it would catch the error and print the line number (in the input file) and single char or the attribute causing the error, so tracing an error with huge data sets  would be less challenging.     Often "line numbers" are challenging in a JSON file, since there are often no line breaks in the file; so maybe it's not easy to do in C++?

FYI, I was already filtering for UTF-8 compliance (when creating the JSON input file), but that did not pass your code; so to get past the issues above I wrote some pre-filters to make sure all basic special chars like "<" or "@" were filtered  out of the JSON input file to avoid fdl errors ; and I had to pre-filter and cast all integers to strings; and it seems to work OK now with all this pre-filtering in place.

Thanks for this code!    

It's still by far the best written and visually  pleasing  FDG  code we have found.

**Comment by godlikemouse** [comment]
Hi unixneo,

If you could include some of the offending JSON that would be great.  It would make it a lot easier to track down the issues and fix it.  Also, did your JSON pass external validation using a validator?

**Comment by unixneo** [comment]
Hey godlikemouse,

Please send you send me your email address, I will zip and email the entire XML file to you.

Cheers!

**Comment by godlikemouse** [comment]
Hi unixneo,

I don't need the whole file, please just post the lines that are responsible for the issue.  Thanks.

**Comment by godlikemouse** [comment]
Hi unixneo,

Also, if you want to upload just a smaller version which has the errors in it that would be very helpful.  Thanks.

**Comment by unixneo** [comment]
Hey godlikemouse,

OK... let's hold off until I rewrite and clean up my code which builds the initial JSON file.  I don't want to take your valuable time debugging C++ code until I have insured myself my back end PHP code which generates the initial JSON file is producing an output free of errors, blanks, etc.

So let me get my house in order first...  !!

**Comment by godlikemouse** [comment]
Sounds good, I'm using a third party JSON encode/decode library that I should be able to turn on more error information (just did a quick glance through it the other day).  Let me know when you're ready to proceed.  Thanks.

**Comment by unixneo** [comment]
Hey godlikemouse,

FYI, since I have experienced that FDGs take a long time to produce a graph when the data set is large, I've created a faster algo for writing graphs, so I'm putting FDG on the back burner for now.

Your excellent code really helped me a lot; and I will always appreciate and be thankful for it.

Due to general performance issues with FDG over large graphs, I have decided to go with a different graph algorithm for my projects.    The results are posted on my blog.

Cheers and Thank you very much again!
